cpp/src/io/orc/stats_enc.cu

```diff
@@ -0,0 +1,380 @@
+  return p;
+}
+
+// Protobuf field encoding for unsigned int
+__device__ inline uint8_t *pb_put_uint(uint8_t *p, uint32_t id, uint64_t v) {
+  p[0] = id * 8 + PB_TYPE_VARINT; // NOTE: Assumes id < 16
+  return pb_encode_uint(p + 1, v);
+}
+
+// Protobuf field encoding for signed int
+__device__ inline uint8_t *pb_put_int(uint8_t *p, uint32_t id, int64_t v) {
+  int64_t s = (v < 0);
+  return pb_put_uint(p, id, (v ^ -s) * 2 + s);
+}
+
+// Protobuf field encoding for 'packed' unsigned int (single value)
+__device__ inline uint8_t *pb_put_packed_uint(uint8_t *p, uint32_t id, uint64_t v) {
+  uint8_t *p2 = pb_encode_uint(p + 2, v);
+  p[0] = id * 8 + PB_TYPE_FIXEDLEN;
+  p[1] = static_cast<uint8_t>(p2 - (p + 2));
+  return p2;
+}
+
+// Protobuf field encoding for binary/string
+__device__ inline uint8_t *pb_put_binary(uint8_t *p, uint32_t id, const void *bytes, uint32_t len) {
+  p[0] = id * 8 + PB_TYPE_FIXEDLEN;
+  p = pb_encode_uint(p + 1, len);
+  memcpy(p, bytes, len);
+  return p + len;
+}
+
+// Protobuf field encoding for 64-bit raw encoding (double)
+__device__ inline uint8_t *pb_put_fixed64(uint8_t *p, uint32_t id, const void *raw64) {
+  p[0] = id * 8 + PB_TYPE_FIXED64;
+  memcpy(p + 1, raw64, 8);
+  return p + 9;
+}
+
+
+
+/**
+ * @brief Encode statistics in ORC protobuf format
+ *
+ * @param[in,out] groups Statistics merge groups
+ * @param[in,out] chunks Statistics data
+ * @param[in] statistics_count Number of statistics buffers
+ *
+ **/
+// blockDim {128,1,1}
+__global__ void __launch_bounds__(128)
+gpuEncodeStatistics(uint8_t *blob_bfr, statistics_merge_group *groups, const statistics_chunk *chunks,
+                    uint32_t statistics_count) {
+  __shared__ __align__(8) stats_state_s state_g[4];
+  uint32_t t = threadIdx.x & 0x1f;
+  uint32_t idx = blockIdx.x * 4 + (threadIdx.x >> 5);
+  stats_state_s * const s = &state_g[threadIdx.x >> 5];
+  if (idx < statistics_count) {
+    if (t < sizeof(statistics_chunk) / sizeof(uint32_t)) {
+      reinterpret_cast<uint32_t *>(&s->ck)[t] = reinterpret_cast<const uint32_t *>(&chunks[idx])[t];
+    }
+    if (t < sizeof(statistics_merge_group) / sizeof(uint32_t)) {
+      reinterpret_cast<uint32_t *>(&s->grp)[t] = reinterpret_cast<uint32_t *>(&groups[idx])[t];
+    }
+  }
+  __syncthreads();
+  if (idx < statistics_count) {
+    if (t < sizeof(stats_column_desc) / sizeof(uint32_t)) {
+      reinterpret_cast<uint32_t *>(&s->col)[t] = reinterpret_cast<const uint32_t *>(s->grp.col)[t];
+    }
+    if (t == 0) {
+      s->base = blob_bfr + s->grp.start_chunk;
+      s->end = blob_bfr + s->grp.start_chunk + s->grp.num_chunks;
+    }
+  }
+  __syncthreads();
+  // Encode and update actual bfr size
+  if (idx < statistics_count && t == 0) {
+    uint8_t *cur = pb_put_uint(s->base, 1, s->ck.non_nulls);
+    uint8_t *fld_start = cur;
+    switch(s->col.stats_dtype) {
+    case dtype_int8:
+    case dtype_int16:
+    case dtype_int32:
+    case dtype_int64:
+      // IntegerStatistics = 2
```

Comment:
Like the one above, this switch has a lot of magic numbers. Please document / clarify.

What does `IntegerStatistics = 2` (and the similar comments in the other cases below) signify?

Reply:
The protobuf field id. Perhaps I'll just paste the ORC protobuf definitions in a big comment section so the reader doesn't have to look at the ORC documentation...